### DISCUSSION TITLE
Is it possible to resolve custom tags remotely?

### DISCUSSION THREAD

**Discussion by jindong-zhang-git** [eemeli/yaml#359] [Q&A]
Hi team.

Due to the architecture of my app, I'd like to define some remote custom tag resolvers, which should be involved when parsing YAMLs, to provide dynamic custom tag definitions.

And the communication between systems uses a JSON based RPC. So only plain datas are supported during communication.

Can I know your opinions about the possibility? Thanks.

**Comment by eemeli** [comment]
The custom tag resolvers are expected to be synchronous, as allowing asynchrony there would effectively require everything up to and including `YAML.parse()` to also be asynchronous.

So if your remote calls are sync, there should be no issue in writing a [custom tag](https://eemeli.org/yaml/#writing-custom-tags) for handling them. OTOH, if they're async, then you're probably better off parsing the source first with `YAML.parseDocument()` and then using a [visitor](https://eemeli.org/yaml/#modifying-nodes) to do your remote resolution.

**Reply by suerta-git** [reply]
Awesome! Looks great to me.

Another 2 questions:
1. Can visitor be an async function? Don't find example in doc.
2. Are AST nodes serializable? Because the communication bases on JSON, so everything must be serializable.

Thank you very much.

**Reply by eemeli** [reply]
1. Huh, good point. Yes, in that you're free to call async code from a visitor and then modify the node later. No, in that you can't return a promise from a visitor. I filed issue #362 for this.
2. If you only need the actual data, then yes, they all have a `toJSON()` method.

**Reply by jindong-zhang-git** [reply]
> 2. If you only need the actual data, then yes, they all have a toJSON() method.

I see but its result only contains "actual data" as you said. But in this scenario, the remote resolver should return AST nodes which contains not only actual data, but also all of metadata, to construct a document as same as the one generated by local parser.

**Reply by eemeli** [reply]
Perhaps this would be easier to understand if you could provide an example of a YAML input and the remotely determined value that it should resolve to?

**Reply by suerta-git** [reply]
Hi @eemeli, this is my another account.

> Perhaps this would be easier to understand if you could provide an example of a YAML input and the remotely determined value that it should resolve to?

If I parse a yaml document using `YAML.parseDocument()` directly, the result will be content nodes `Scalar | YAMLMap | YAMLSeq` which can contain "metadata" like `comment`, `range`, `tag` in addition to actual data.

So I expect the remote resolvers also return such content nodes, then I can insert them into parsed yaml document.

But I can't just serialize content nodes using `JSON.stringify()` because its result looses nodes' "metadata". Is there a simple way that I can transport those data via JSON and then restore them to `Scalar | YAMLMap | YAMLSeq` nodes in consumer?

**Reply by eemeli** [reply]
Then you're talking about a different operation than "tag resolution", which really should only consider the actual scalar or collection value of a node.

Again, an actual example would really help.

If you do need to transport a representation of a node to a remote processor without losing any of the metadata, then why not send it as YAML?
